go/fix-chain/fixer.go

```diff
@@ -0,0 +1,282 @@
+package fix_chain
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"github.com/google/certificate-transparency/go/x509"
+	"log"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const HashSize = sha256.Size
+
+func Hash(s *x509.Certificate) [HashSize]byte {
+	return sha256.Sum256(s.Raw)
+}
+
+func HashChain(ch []*x509.Certificate) (r [HashSize]byte) {
+	h := sha256.New()
+	for _, c := range ch {
+		// FIXME: surely there's an easier way?
+		h2 := make([]byte, HashSize)
+		h3 := Hash(c)
+		copy(h2, h3[:])
+		h.Write(h2)
+	}
+	h2 := h.Sum([]byte{})
+	copy(r[:], h2)
+	return
+}
+
+type Bag struct {
+	certs []*x509.Certificate
+}
+
+func (b Bag) Len() int { return len(b.certs) }
+func (b Bag) Less(i, j int) bool {
+	ci := b.certs[i].Raw
+	cj := b.certs[j].Raw
+	if len(ci) < len(cj) {
```

Comment:
These next 6 lines could be condensed to:
```go
if len(ci) != len(cj) {
  return len(ci) < len(cj)
}
```

Up to you whether you think that's a bad space/readability tradeoff.